src/ORM/DataObjectSchema.php

```diff
@@ -666,67 +666,90 @@ protected function buildCustomDatabaseIndexes($class)
-                        $shouldAddToComposite = false;
-                        continue;
-                    }
-                    // Add indexes as appropriate
-                    if ($shouldAddToSingle) {
-                        $indexes[$column] = [
-                            'type' => 'index',
-                            'columns' => [$column],
-                        ];
-                    }
+        $sort = $this->normaliseSort($sort);
+        foreach ($sort as $value) {
+            try {
+                list ($table, $column, $dir) = $this->parseSortColumn(trim($value ?? ''));
+                $table = trim($table ?? '', '"');
+                $column = trim($column ?? '', '"');
+                // Skip and stop grabbing composite columns if the sort column is on a different table
+                if ($table && strtolower($table ?? '') !== strtolower($tableName)) {
+                    $shouldAddToComposite = false;
+                    continue;
+                }
+                // ID is always the primary key, so we don't need a new index for it.
+                if ($column === 'ID') {
                     if ($shouldAddToComposite) {
+                        // We still need to include it in the composite index if it's part-way through
                         $compositeCols[$column] = "$column $dir";
                     }
-                } catch (InvalidArgumentException $e) {
+                    continue;
+                }
+                // Skip and stop grabbing composite columns if this isn't a column in the database.
+                if (!in_array($column, $fieldNames)) {
+                    $shouldAddToComposite = false;
+                    continue;
+                }
+                // Add indexes as appropriate
+                if ($shouldAddToSingle) {
+                    $indexes[$column] = [
+                        'type' => 'index',
+                        'columns' => [$column],
+                    ];
                 }
+                if ($shouldAddToComposite) {
+                    $compositeCols[$column] = "$column $dir";
+                }
+            } catch (InvalidArgumentException $e) {
             }
+        }
 
-            // If ID is last, we can omit it since that gets implicitly added to all indexes
-            if (array_key_last($compositeCols) === 'ID') {
-                unset($compositeCols['ID']);
-            }
-            // Add a composite index if we either didn't already add a single column or have multiple columns.
-            if (!empty($compositeCols) && (!$shouldAddToSingle || count($compositeCols) > 1)) {
-                $indexes['default_sort_composite'] = [
-                    'type' => 'index',
-                    'columns' => array_values($compositeCols),
-                ];
-            }
+        // If ID is last, we can omit it since that gets implicitly added to all indexes
+        if (array_key_last($compositeCols) === 'ID') {
```

Comment:
does it matter if ID is latst? What if it's in the middle of the array for whatever reason, we'd still want to remove it? Use array_key_exists() instead?

Reply:
This code already existed. I'm not inclined to change it in this PR. But yes, it being last does matter - if it's last, it's already implicitly added so adding it _again_ is just a waste.
If it's in the middle, that's _not_ already implicitly there and we can't skip columns, so we need to explicitly include it.